kani-compiler/src/codegen_boogie/overrides/hooks.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module contains various codegen hooks for functions that require special handling.
+//!
+//! E.g.: Functions in the Kani library that generate assumptions or symbolic variables.
+//!
+
+use crate::codegen_boogie::BoogieCtx;
+use boogie_ast::boogie_program::{Expr, Stmt};
+use rustc_middle::mir::{BasicBlock, Place};
+use rustc_middle::ty::{Instance, TyCtxt};
+use rustc_span::Span;
+use std::rc::Rc;
+use tracing::debug;
+
+pub trait BoogieHook<'tcx> {
```

Comment:
😕 You might want to simplify this a bit later. One would hope that we don't replicate the hooks logic. Our hooks currently implement two different things: kani intrinsics (`kani::assert`, `kani::assume`, `kani::cover`) and function abstraction.

I would suggest that you handle them separately (as we should eventually in the GotoC too). In your current implementation, you only care about intrinsics, so you can only implement the first for now. Hopefully we can rely on stubbing for the second, so you won't need to add anything extra in the backend for that.

For intrinsics, you should basically just check if the function has any diagnostic item, and do a table lookup to see if that is an intrinsic.

Reply:
Makes sense. I removed the hooks module, and created a new module for handling Kani intrinsics. I placed it for now under the `codegen_boogie` module, but we should move it up to `kani_middle` at some point.